### PR TITLE
Convert to Headless Program

### PR DESCRIPTION
**Pull request by werkshy**
- Drop support for dynosaur-rails
- Only use YAML for config
- Create pluggable error handlers and stats handlers
- Add cronut ping support
- Add specs

- [x] Create harrys-dynosaur heroku app
- [x] Set up cronut job
- [x] Update docs and sample config
- [x] Set dashboard to use new metric name

### CODE REVIEW

File: dynosaur.gemspec [hunk 1]
```diff
@@ -6,8 +6,8 @@ require 'dynosaur/version'
 Gem::Specification.new do |spec|
   spec.name          = "dynosaur"
   spec.version       = Dynosaur::VERSION
-  spec.authors       = ["Andy O'Neill", "Daniel Schwartz"]
-  spec.email         = ["aoneill@harrys.com",  'daniel@harrys.com']
+  spec.authors       = ["Andy O'Neill", "Pierre Jambet", "Daniel Schwartz"]
+  spec.email         = ["aoneill@harrys.com",  'pjambet@harrys.com']
```
Comment (pjambet) [inline dynosaur.gemspec]: not my email!

File: bin/dynosaur [hunk 2]
```diff
@@ -2,25 +2,25 @@
 
 require 'rubygems'
 
-$LOAD_PATH.unshift File.join(File.dirname(__FILE__), *%w[.. lib])
+#$LOAD_PATH.unshift File.join(File.dirname(__FILE__), *%w[.. lib])
```
Comment (pjambet) [inline bin/dynosaur]: why is this commented?

File: lib/dynosaur/addons.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      # @params compare_field Field used to compare plans
+      #
+      def plans_for_addon(addon, compare_field = 'tier')
+        return all[addon].map { |plan|
+          AddonPlan.new(plan, compare_field)
+        }
+      end
+
+    end
+  end
```
Comment (pjambet) [inline lib/dynosaur/addons.rb]: :+1:

File: lib/dynosaur/autoscaler.rb [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+      end
+    end
+
+    def global_config(scaler_config)
+      # Get the values from the global config
+      if scaler_config.nil?
+        raise StandardError.new "Please include a 'scaler' block in the config"
+      end
+      @dry_run = scaler_config.fetch("dry_run", @dry_run)
+      # @stats = scaler_config.fetch("stats", @stats)
```
Comment (pjambet) [inline lib/dynosaur/autoscaler.rb]: do we still need this?

File: lib/dynosaur/controllers/papertrail_controller_plugin.rb [hunk 5]
```diff
@@ -10,11 +10,11 @@ def initialize(config)
 
         @min_resource = Dynosaur::Addons.plans_for_addon('papertrail').find {|plan| plan['name'] == min_resource_name }
         if @min_resource.nil?
-          raise "Min resource not found with name #{min_resource_name}"
+          raise StandardError.new "Min resource not found with name #{min_resource_name}"
```
Comment (pjambet) [inline lib/dynosaur/controllers/papertrail_controller_plugin.rb]: doesn't `raise` create a `StandardError` by default? or are you just doing that for clarity?

File: lib/dynosaur/error_handler/ses.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        msg += exception.backtrace.join("<br>")
+        msg += "</div>"
+
+        @ses.send_email(
+          from: @from,
+          to: @to,
+          subject: "Dynosaur Error",
+          body_html: msg)
+    end
+  end
```
Comment (pjambet) [inline lib/dynosaur/error_handler/ses.rb]: :+1:

File: lib/dynosaur/heroku_dyno_manager.rb [hunk 7]
```diff
@@ -2,14 +2,14 @@ module Dynosaur
...
-      return @state["dynos"]
+      @state = @heroku_platform_api.formation.info(@app_name, 'web')
+      return @state["quantity"]
     end
 
     def set(value)
       puts "Setting current dynos to #{value}"
       if !@dry_run
-        @heroku.post_ps_scale(@app_name, 'web', value)
+        @heroku_platform_api.formation.update(@app_name, 'web', {quantity: value})
```
Comment (pjambet) [inline lib/dynosaur/heroku_dyno_manager.rb]: Finally!

File: lib/dynosaur/inputs/abstract_input_plugin.rb [hunk 8]
```diff
@@ -64,10 +69,8 @@ def min_recent_values
         return @recent.min
       end
 
-      def get_status
+      def health
         now = Time.now
-        value = get_value
-        estimate = estimated_resources  # minor race condition, but only matters for logging
         health = "OK"
```
Comment (pjambet) [inline lib/dynosaur/inputs/abstract_input_plugin.rb]: Using the same name for the variable and the method sounds like some potential future headache

File: lib/dynosaur/inputs/google_analytics_input_plugin.rb [hunk 9]
```diff
@@ -69,7 +59,9 @@ def init_api
         @client.authorization = nil
         if @keytext  # load key from string
           puts "Loading key from PEM text"
-          @key = OpenSSL::PKey::RSA.new(@keytext)
+          # When we load from YAML we replaced newlines with pipes, let's undo
+          # that
+          @key = OpenSSL::PKey::RSA.new(@keytext.gsub("|", "\n"))
```
Comment (pjambet) [inline lib/dynosaur/inputs/google_analytics_input_plugin.rb]: why do we convert newlines with pipes in the first place?

File: lib/dynosaur/success/cronut.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      if ping.status != 200
+        puts "ERROR: CRONUT RETURNED #{ping.status}"
+      end
+      @last_pinged = Time.now
+    rescue StandardError => e
+      puts "ERROR FROM CRONUT: #{e.inspect}"
+      puts e.backtrace.join("\n")
+    end
+  end
+end
```
Comment (pjambet) [inline lib/dynosaur/success/cronut.rb]: Awesome

**Comment by pjambet** [comment]
Looks great!

File: lib/dynosaur/inputs/google_analytics_input_plugin.rb (hunk 9 above)
Comment (werkshy) [inline lib/dynosaur/inputs/google_analytics_input_plugin.rb]: It breaks YAML when you insert it with newlines and not properly indented. Ugh.

File: bin/dynosaur (hunk 2 above)
Comment (werkshy) [inline bin/dynosaur]: I'll rip it out

File: lib/dynosaur/autoscaler.rb (hunk 4 above)
Comment (werkshy) [inline lib/dynosaur/autoscaler.rb]: will kill

File: lib/dynosaur/controllers/papertrail_controller_plugin.rb (hunk 5 above)
Comment (werkshy) [inline lib/dynosaur/controllers/papertrail_controller_plugin.rb]: Ohhhhh!

File: lib/dynosaur/inputs/abstract_input_plugin.rb (hunk 8 above)
Comment (werkshy) [inline lib/dynosaur/inputs/abstract_input_plugin.rb]: indeed

**Pull request merged**